arch/arm/src/stm32wb/stm32wb_i2c.c

```diff
@@ -0,0 +1,2650 @@
+
+static inline void
+stm32wb_i2c_set_bytes_to_transfer(struct stm32wb_i2c_priv_s *priv,
+                                  uint8_t n_bytes)
+{
+  stm32wb_i2c_modifyreg32(priv, STM32WB_I2C_CR2_OFFSET, I2C_CR2_NBYTES_MASK,
+                          (n_bytes << I2C_CR2_NBYTES_SHIFT));
+}
+
+/****************************************************************************
+ * Name: stm32wb_i2c_set_write_transfer_dir
+ *
+ * Description:
+ *
+ ****************************************************************************/
+
+static inline void
+stm32wb_i2c_set_write_transfer_dir(struct stm32wb_i2c_priv_s *priv)
+{
+  stm32wb_i2c_modifyreg32(priv, STM32WB_I2C_CR2_OFFSET, I2C_CR2_RD_WRN, 0);
+}
+
+/****************************************************************************
+ * Name: stm32wb_i2c_set_read_transfer_dir
+ *
+ * Description:
+ *
+ ****************************************************************************/
+
+static inline void
+stm32wb_i2c_set_read_transfer_dir(struct stm32wb_i2c_priv_s *priv)
+{
+  stm32wb_i2c_modifyreg32(priv, STM32WB_I2C_CR2_OFFSET,
+                          0, I2C_CR2_RD_WRN);
+}
+
+/****************************************************************************
+ * Name: stm32wb_i2c_enable_reload
+ *
+ * Description:
+ *
+ ****************************************************************************/
+
+static inline void
+stm32wb_i2c_enable_reload(struct stm32wb_i2c_priv_s *priv)
+{
+  stm32wb_i2c_modifyreg32(priv, STM32WB_I2C_CR2_OFFSET,
+                          0, I2C_CR2_RELOAD);
+}
+
+/****************************************************************************
+ * Name: stm32wb_i2c_disable_reload
+ *
+ * Description:
+ *
+ ****************************************************************************/
+
+static inline void
+stm32wb_i2c_disable_reload(struct stm32wb_i2c_priv_s *priv)
+{
+  stm32wb_i2c_modifyreg32(priv, STM32WB_I2C_CR2_OFFSET,
+                          I2C_CR2_RELOAD, 0);
+}
+
+/****************************************************************************
+ * Name: stm32wb_i2c_sem_waitstop
+ *
+ * Description:
+ *   Wait for a STOP to complete
+ *
+ ****************************************************************************/
+
+static inline
+void stm32wb_i2c_sem_waitstop(struct stm32wb_i2c_priv_s *priv)
+{
+  clock_t start;
+  clock_t elapsed;
+  clock_t timeout;
+  uint32_t cr;
+  uint32_t sr;
+
+  /* Select a timeout */
+
+#ifdef CONFIG_STM32WB_I2C_DYNTIMEO
+  timeout = USEC2TICK(CONFIG_STM32WB_I2C_DYNTIMEO_STARTSTOP);
+#else
+  timeout = CONFIG_STM32WB_I2CTIMEOTICKS;
+#endif
+
+  /* Wait as stop might still be in progress */
+
+  start = clock_systime_ticks();
+  do
+    {
+      /* Calculate the elapsed time */
+
+      elapsed = clock_systime_ticks() - start;
+
+      /* Check for STOP condition */
+
+      cr = stm32wb_i2c_getreg32(priv, STM32WB_I2C_CR2_OFFSET);
+      if ((cr & I2C_CR2_STOP) == 0)
+        {
+          return;
```

Comment:
Maybe it could reported using i2cerr()

Reply:
Actually this `return` is a successful flow.